src/wayland/output/mod.rs

```diff
@@ -339,82 +355,95 @@ impl Output {
         }
 
         for output in &inner.instances {
             if let Some(mode) = new_mode {
-                output.mode(flags, mode.size.w, mode.size.h, mode.refresh);
+                output.mode(dh, flags, mode.size.w, mode.size.h, mode.refresh);
             }
             if new_transform.is_some() || new_location.is_some() {
-                inner.send_geometry(output);
+                inner.send_geometry_to(dh, output);
             }
             if let Some(scale) = new_scale {
-                if output.as_ref().version() >= 2 {
-                    output.scale(scale);
+                if output.version() >= 2 {
+                    output.scale(dh, scale);
                 }
             }
-            if output.as_ref().version() >= 2 {
-                output.done();
+            if output.version() >= 2 {
+                output.done(dh);
             }
         }
     }
 
     /// Check is given [`wl_output`](WlOutput) instance is managed by this [`Output`].
     pub fn owns(&self, output: &WlOutput) -> bool {
-        self.inner
+        self.data
+            .inner
             .0
             .lock()
             .unwrap()
             .instances
             .iter()
-            .any(|o| o.as_ref().equals(output.as_ref()))
+            .any(|o| o.id() == output.id())
     }
 
     /// This function allows to run a [FnMut] on every
     /// [WlOutput] matching the same [Client] as provided
-    pub fn with_client_outputs<F>(&self, client: Client, f: F)
+    pub fn with_client_outputs<F>(&self, dh: &mut DisplayHandle<'_>, client: &Client, mut f: F)
     where
-        F: FnMut(&WlOutput),
+        F: FnMut(&mut DisplayHandle<'_>, &WlOutput),
     {
-        self.inner
+        let list: Vec<_> = self
+            .data
+            .inner
             .0
             .lock()
             .unwrap()
             .instances
             .iter()
-            .filter(|output| match output.as_ref().client() {
-                Some(output_client) => output_client.equals(&client),
-                None => false,
+            .filter(|output| {
+                dh.get_client(output.id())
+                    .map(|output_client| &output_client == client)
+                    .unwrap_or(false)
             })
-            .for_each(f)
+            .cloned()
+            .collect();
+
+        for o in list {
+            f(dh, &o);
+        }
     }
 
     /// Sends `wl_surface.enter` for the provided surface
     /// with the matching client output
-    pub fn enter(&self, surface: &wl_surface::WlSurface) {
-        if let Some(client) = surface.as_ref().client() {
-            self.with_client_outputs(client, |output| surface.enter(output))
+    pub fn enter(&self, dh: &mut DisplayHandle<'_>, surface: &wl_surface::WlSurface) {
+        let client = dh.get_client(surface.id()).ok();
+
+        if let Some(client) = client {
+            self.with_client_outputs(dh, &client, |dh, output| surface.enter(dh, output))
         }
```

Comment:
If we're going to just match it using `if let`, what's the point of using `.ok()` here?